docs/architecture/architecture.md

```diff
@@ -609,6 +614,117 @@ rootfs by calling `mount(8)` within the counter.  If the devicemapper block devi
+However, due to the fact that Kubernetes execution units are sets of containers (also
+known as pods) rather than single containers, the Clear Containers runtime needs to
+get extra information to seamlessly integrate with Kubernetes.
+
+### Problem statement
+
+[Kubernetes](https://github.com/kubernetes/kubernetes/) execution unit is a `Pod`,
+i.e. a set of containers sharing the same resources and networking, mount and PID
+namespaces. A pod specification is a set of containers associated to a list of
+constraints (namespaces, cgroups, hardware resources, security contexts, etc...).
+By default the Clear Containers runtime associates each container with a dedicated
+virtual machine, but that does not work really well with pods as letting a set of
+VMs share the same namespaces and resources can be very complex to implement. For
+example allowing 2 virtual machines to share the same networking or PID namespace
+can be very challenging.
+
+By default the Clear Containers runtime creates a virtual machine for each container,
+but with Kubernetes it will create and launch one virtual machine per pod. Then each
+container within a pod is simply a container inside the VM boundary.
+
+So the challenge with Clear Containers when trying to work as a `CRI-O` runtime is
+to know when to create a full virtual machine (for pods) and when to create a new
+container inside a previously created VM. In both cases it will get called with very
+similar arguments, so it needs `CRI-O` help to be able to distinguish between a pod
+creation from a container one.
+
+### OCI annotations
+
+In order for the Clear Containers runtime (Or any VM based OCI compatible runtime) to
+be able to understand if it needs to create a full virtual machine or if it has to create
+a new container inside an existing pod's VM context, `CRI-O` adds specific annotations
+to the OCI configuration file (`config.json`) passed down to the OCI compatible runtimes
+it relies on.
+
+Before calling its runtime, `CRI-O` will always add a `io.kubernetes.cri-o.ContainerType`
+annotation to the `config.json` OCI configuration files it produces from the kubelet CRI
+request. `io.kubernetes.cri-o.ContainerType` can either be set to `sandbox` or `container`.
+Clear Containers will then use this annotation to decide if it needs to respectively
+create a virtual machine or a container inside a virtual machine associated with a
+Kubernetes pod:
+
+```Go
+	containerType, err := ociSpec.ContainerType()
+	if err != nil {
+		return err
+	}
+
+	switch containerType {
+	case vc.PodSandbox:
+		process, err = createPod(ociSpec, runtimeConfig, containerID, bundlePath, console, disableOutput)
+		if err != nil {
+			return err
+		}
+	case vc.PodContainer:
+		process, err = createContainer(ociSpec, containerID, bundlePath, console, disableOutput)
+		if err != nil {
+			return err
+		}
+	}
+
+```
+
+### Generalization
+
+With Clear Containers 2.1, we implemented the initial Kubernetes/`CRI-O` support by choosing
```

Comment:
- Missing spaces either side of the slash.
- I'd write this impersonally maybe ("Clear Containers 2.1 implemented the initial ...").